src/lib.rs

```diff
@@ -255,3 +259,85 @@ impl Defaults {
         }
     }
 }
+
+#[derive(Debug)]
+pub enum Usage {
+    Read,
+    Write,
+    ReadWrite,
+}
+
+impl Usage {
+    fn parse(tree: &Element) -> Usage {
+        let text = try!(tree.text.as_ref());
+
+        match &text[..] {
+            "read" => Usage::Read,
+            "write" => Usage::Write,
+            "read-write" => Usage::ReadWrite,
+            _ => panic!("unknown usage variant: {}", text),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EnumeratedValues {
+    pub name: Option<String>,
+    pub usage: Option<Usage>,
+    pub derived_from: Option<String>,
+    pub values: Vec<EnumeratedValue>,
+}
+
+impl EnumeratedValues {
+    fn parse(tree: &Element) -> EnumeratedValues {
+        assert_eq!(tree.name, "enumeratedValues");
+
+        EnumeratedValues {
+            name: tree.get_child_text("name"),
+            usage: tree.get_child("usage").map(Usage::parse),
+            derived_from: tree.attributes.get(&"derivedFrom".to_owned())
+                .map(|s| s.to_owned()),
+            values: tree.children.iter()
+                .filter_map(EnumeratedValue::parse)
+                .collect(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Choice {
+    value: Option<u32>,
+    is_default: Option<bool>,
+}
+
+impl Choice {
+    fn parse(tree: &Element) -> Choice {
+        Choice {
+            value: try!(tree.get_child_text("value")).parse().ok(),
```

Comment:
This looks wrong. 'value: Option<u32>' indicates that the field may or may not appear but 'try!(tree.get_child_text("value"))' states that the field *must always* appear (because our try! is equivalent to unwrapping). This should be 'tree.get_child_text("value").map(|s| try!(s.parse()))'.